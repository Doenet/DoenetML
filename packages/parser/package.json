{
    "name": "@doenet/parser",
    "type": "module",
    "description": "Parse DoenetML",
    "version": "*",
    "license": "AGPL-3.0-or-later",
    "homepage": "https://github.com/Doenet/DoenetML#readme",
    "private": false,
    "repository": "github:Doenet/DoenetML",
    "files": [
        "/dist"
    ],
    "exports": {
        ".": {
            "import": "./dist/index.js"
        }
    },
    "scripts": {
        "dev": "vite",
        "watch": "vite build --watch",
        "build": "wireit",
        "test": "vitest",
        "compile_grammar": "npx lezer-generator --output src/generated-assets/lezer-doenet.ts src/doenet.grammar"
    },
    "wireit": {
        "build": {
            "command": "vite build",
            "files": [
                "src/**/*.ts",
                "tsconfig.json",
                "vite.config.ts"
            ],
            "output": [
                "dist/**/*.js",
                "dist/**/*.d.ts",
                "dist/**/*.json"
            ],
            "dependencies": [
                "../static-assets:build"
            ]
        }
    },
    "dependencies": {
        "@lezer/common": "^1.2.1",
        "@lezer/generator": "^1.7.0",
        "@lezer/lr": "^1.4.1",
        "@types/xast": "^2.0.4",
        "peggy": "^4.0.2",
        "prettier-plugin-pegjs": "^2.0.2",
        "unified": "^11.0.4",
        "xast-util-to-xml": "^4.0.0"
    },
    "devDependencies": {
        "react18-json-view": "^0.2.8"
    }
}
