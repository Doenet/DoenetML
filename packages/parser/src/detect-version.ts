import { lezerToDast } from "./lezer-to-dast";
import { DastElement } from "./types";

/**
 * If doenetML consists of a single `<document>` component with,
 * for example, `xmlns="https://doenet.org/spec/doenetml/v0.7.0-alpha44"` set,
 * attempt to extract the DoenetML version from that prop.
 * Returns the version if found, else returns null.
 */
export function detectVersionFromDoenetML(
    doenetML: string,
): { version: string; position: { begin: number; end: number } } | null {
    const parsed = lezerToDast(doenetML);
    // Find the first document child and look for the xmlns attribute
    const documentElement = parsed.children.find(
        (e) => e.type === "element" && e.name === "document",
    ) as DastElement;
    if (!documentElement || !("xmlns" in documentElement.attributes)) {
        return null;
    }
    const xmlnsValue = documentElement.attributes["xmlns"];
    if (
        xmlnsValue.type !== "attribute" ||
        xmlnsValue.children.length !== 1 ||
        xmlnsValue.children[0].type !== "text"
    ) {
        return null;
    }
    // Should now be a URL like https://doenet.org/spec/doenetml/v0.7.0-alpha44
    const xmlnsTextNode = xmlnsValue.children[0];
    const xmlnsText = xmlnsValue.children[0].value;

    const match = xmlnsText.match(
        /^https:\/\/doenet\.org\/spec\/doenetml\/v([0-9a-zA-Z\.\-]+)$/,
    );

    const documentTag = match ? match[0] : null;
    const versionString = match ? match[1] : null;
    if (match && documentTag && versionString != null) {
        // Calculate the offset of the match
        const offsetStart =
            (match.index || 0) + documentTag.indexOf(versionString);
        const offsetEnd = offsetStart + versionString.length;
        return {
            version: versionString,
            position: {
                begin: xmlnsTextNode.position?.start.offset! + offsetStart,
                end: xmlnsTextNode.position?.start.offset! + offsetEnd + 1,
            },
        };
    }

    return null;
}
