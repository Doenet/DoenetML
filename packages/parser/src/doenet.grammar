/*
 * Modified version of lezer-xml that accepts more html-like identifiers
 */
@top Document { (entity | DoctypeDecl)+ }

entity {
  Text |
  EntityReference |
  CharacterReference |
  Ampersand |
  Cdata |
  Element |
  Comment |
  ProcessingInst |
  MismatchedCloseTag |
  incompleteStartCloseTag
}

Element {
  OpenTag entity* (CloseTag | MissingCloseTag) |
  SelfClosingTag
}

@skip { space } {
  OpenTag[closedBy="CloseTag MissingCloseTag"] {
    StartTag TagName Attribute* EndTag
  }

  SelfClosingTag {
    StartTag TagName Attribute* SelfCloseEndTag
  }

  CloseTag[openedBy=OpenTag] {
    StartCloseTag TagName EndTag
  }

  MismatchedCloseTag {
    mismatchedStartCloseTag TagName EndTag
  }

  Attribute {
    AttributeName (Is AttributeValue)?
  }
}

AttributeValue {
  "\"" (doubleQuoteAttributeContent | EntityReference | CharacterReference | Ampersand)* "\"" |
  "'" (singleQuoteAttributeContent | EntityReference | CharacterReference | Ampersand)* "'"
}

Comment { "<!--" commentContent* "-->" }

ProcessingInst { "<?" piContent* "?>" }

Cdata { cdataStart cdataContent* "]]>" }

@context elementContext from "../tokens"

@external tokens startTag from "../tokens" {
  StartTag[closedBy="SelfCloseEndTag EndTag"]
  StartCloseTag
  MissingCloseTag
  mismatchedStartCloseTag[@name=StartCloseTag]
  incompleteStartCloseTag[@name=StartCloseTag]
}

@external tokens commentContent from "../tokens" { commentContent }
@external tokens piContent from "../tokens" { piContent }
@external tokens cdataContent from "../tokens" { cdataContent }
@external tokens textLessThan from "../tokens" { textLessThan }

@tokens {
  EndTag[openedBy="StartTag StartCloseTag"] { ">" }

  SelfCloseEndTag[openedBy=StartTag] { "/>" }

  nameStart {
    ":" | $[a-zA-Z] | "_" |
    $[\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D] |
    $[\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\u{10000}-\u{EFFFF}]
  }

  nameChar {
    nameStart | "-" | "." | $[0-9] | $[\u00B7\u0300-\u036F\u203F-\u2040]
  }

  identifier { nameStart nameChar* }

  TagName { identifier }

  AttributeName { identifier }

  singleQuoteAttributeContent { !['&]+ }
  doubleQuoteAttributeContent { !["&]+ }

  Is { "=" }

  // See https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-references
  EntityReference { "&" identifier ";" }

  CharacterReference { ("&#x" $[0-9a-fA-F]+ ";" | "&#" $[0-9]+ ";") }

  Ampersand { "&" }

  @precedence { CharacterReference, EntityReference, Ampersand }

  Text { ("<" space | "<=" | ![<&])+ }

  DoctypeDecl { "<!" ("doctype" | "DOCTYPE") ![>]* ">" }

  cdataStart { "<![" ("cdata" | "CDATA") "[" }

  space { (" " | "\t" | "\r" | "\n")+ }
}

//@external propSource xmlHighlighting from "./highlight"

@detectDelim
