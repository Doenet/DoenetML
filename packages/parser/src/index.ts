export * from "./parser";
export * from "./extract-dast-errors";
