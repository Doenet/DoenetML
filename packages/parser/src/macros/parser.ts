// This file needs to be here because typescript does not know how to use the transpiler
// to directly load Pegjs grammars.
// @ts-nocheck
import _MacroParser from "./macros.peggy";
import { ParseFunction } from "./types";

type PegParser = {
    parse: ParseFunction;
    SyntaxError: (
        message: string,
        expected: string,
        found: unknown,
        location: unknown,
    ) => unknown;
};

const MacroParser = _MacroParser as PegParser;

export { MacroParser };
