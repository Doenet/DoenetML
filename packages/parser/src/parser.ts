import type { SyntaxNode, TreeCursor } from "@lezer/common";
import { parser } from "./generated-assets/lezer-doenet";

// Re-export parser for CodeMirror instances
export { parser };

type AttrRange = Record<
    string,
    { attrBegin: number; attrEnd: number; begin: number; end: number }
>;
export type Element = {
    componentType: string;
    props: Record<string, string | boolean>;
    children: Node[];
    attributeRanges?: AttrRange;
    state?: { message: string };
    doenetMLrange?: {
        begin?: number;
        end?: number;
        openBegin?: number;
        openEnd?: number;
        closeBegin?: number;
        closeEnd?: number;
        selfCloseBegin?: number;
        selfCloseEnd?: number;
    };
    doenetAttributes?: {
        createNameFromComponentType: string;
    };
};

type DummyElement = {
    componentType: string;
    state: { text: string };
    doenetMLrange: { begin: number; end: number };
};

export type Node = Element | DummyElement | string;

export type ParseError = {
    message: string;
    doenetMLrange: { begin: number; end: number };
};

/**
 *  takes in a string an outputs a TreeCursor
 * @param {string} inText
 * @returns {TreeCursor}
 */
export function parse(inText: string) {
    return parser.parse(inText).cursor();
}
/**
 * parse string and output a convenient to use object.
 * ignores macros.
 * @param {string} inText
 */
export function parseAndCompile(inText: string) {
    let errors: ParseError[] = [];

    function compileElement(cursor: TreeCursor) {
        if (cursor.name !== "Element") {
            throw Error("compileElement() called on a non-Element");
        }
        cursor.firstChild();
        // @ts-ignore
        if (cursor.name === "OpenTag") {
            //skip the start tag node
            cursor.firstChild();
            cursor.nextSibling();
            let tagName = inText.substring(cursor.from, cursor.to);
            let adjustedTagName = tagName;
            let adjustedRange = null;

            let tagOpenBegin = cursor.from;

            let message;

            if (tagName[0] === "_") {
                message = `Invalid component type: <${tagName}>.`;
                errors.push({
                    message,
                    doenetMLrange: { begin: tagOpenBegin, end: cursor.to },
                });
                adjustedTagName = "_error";
            }

            let attrs: Element["props"] = {};
            let attrRanges: AttrRange = {};
            while (cursor.nextSibling()) {
                if (cursor.name === "EndTag") {
                    continue;
                }
                //All of the siblings must b.name Attributes, but we're checking just in case the grammar changes
                if (cursor.name !== "Attribute") {
                    // let errorBegin = cursor.from;
                    let errorEnd = cursor.to;
                    // console.error(cursor);
                    // console.error(showCursor(cursor));
                    // console.error(cursor.name);
                    // eslint-disable-next-line no-empty
                    // while (cursor.parent()) {}

                    let textOfError = inText
                        .slice(tagOpenBegin - 1, errorEnd)
                        .trimEnd();

                    errorEnd = tagOpenBegin + textOfError.length - 1;

                    message = `Invalid DoenetML. Error in opening <${tagName}> tag.  Found ${textOfError}`;

                    errors.push({
                        message,
                        doenetMLrange: { begin: tagOpenBegin, end: errorEnd },
                    });
                    adjustedTagName = "_error";
                    adjustedRange = { begin: tagOpenBegin, end: errorEnd };

                    break;
                }

                //Attributes always have exactly two children, an AttributeName and an Attribute Value
                //We scrape the content of both from the in string and add them to the attribute array here
                cursor.firstChild();
                let attrName = inText.substring(cursor.from, cursor.to);
                let beginAttributeInd = cursor.from + 1;
                //skip the name and equals sign
                if (cursor.nextSibling() === false) {
                    if (attrName in attrs) {
                        message = `Duplicate attribute ${attrName}.`;
                        errors.push({
                            message,
                            doenetMLrange: {
                                begin: beginAttributeInd,
                                end: cursor.to,
                            },
                        });
                        adjustedTagName = "_error";
                    } else {
                        attrs[attrName] = true;
                        attrRanges[attrName] = {
                            attrBegin: beginAttributeInd,
                            attrEnd: cursor.to,
                            begin: beginAttributeInd,
                            end: cursor.to,
                        };
                    }
                } else {
                    cursor.nextSibling();
                    //boundry fuddling to ignore the quotes
                    let attrValue = inText.substring(
                        cursor.from + 1,
                        cursor.to - 1,
                    );

                    if (attrName in attrs) {
                        message = `Duplicate attribute ${attrName}.`;
                        errors.push({
                            message,
                            doenetMLrange: {
                                begin: beginAttributeInd,
                                end: cursor.to,
                            },
                        });
                        adjustedTagName = "_error";
                    } else {
                        attrs[attrName] = attrValue;
                        attrRanges[attrName] = {
                            attrBegin: beginAttributeInd,
                            attrEnd: cursor.to,
                            begin: cursor.from + 2,
                            end: cursor.to - 1,
                        };
                    }
                }
                //move out of Attribute to maintain loop invariant
                cursor.parent();
            }

            //get back to the level of OpenTag in order to parse tag body
            cursor.parent();

            let tagOpenEnd = cursor.to;

            let element: Element = {
                componentType: adjustedTagName,
                props: {},
                children: [],
            };

            if (adjustedTagName === "_error") {
                element.state = { message: message || "" };
                element.doenetAttributes = {
                    createNameFromComponentType: tagName,
                };
                if (attrs.name) {
                    element.props = { name: attrs.name };
                }
            } else {
                element.props = { ...attrs };
                element.attributeRanges = { ...attrRanges };
            }

            // now we go through all of the other non-terminals in this row until we get to the closing tag,
            // adding the compiled version of each non-terminal to the children section of the object we're going to return
            // for the time being we're just going to handle 2 cases:
            // the text case, in which case we'll just push a string into the children,
            // and the element case, in which case we recurse

            // Corresponds to the entity non-terminal in the grammar
            while (cursor.nextSibling()) {
                if (
                    cursor.name === "Text" ||
                    cursor.name === "Ampersand" ||
                    cursor.name === "EntityReference" ||
                    cursor.name === "CharacterReference"
                ) {
                    let txt = inText.substring(cursor.from, cursor.to);
                    if (txt !== "") {
                        element.children.push(txt);
                    }
                } else if (cursor.name === "Element") {
                    element.children.push(compileElement(cursor.node.cursor()));
                } else if (cursor.name === "CloseTag") {
                    // Will always be the matching tag (and the last tag in the list)
                    break;
                } else if (cursor.name === "Comment") {
                    // return a comment that will be ignored,
                    // but need it to calculate doenetMLrange
                    // from any strings that follow it (needed in particular for macros)
                    element.children.push({
                        componentType: "_comment",
                        state: {
                            text: inText.substring(cursor.from, cursor.to),
                        },
                        doenetMLrange: {
                            begin: cursor.from,
                            end: cursor.to,
                        },
                    });
                    //ignore comments
                    continue;
                } else if (cursor.name === "MismatchedCloseTag") {
                    let message = `Invalid DoenetML. Mismatched closing tag.  Expected </${tagName}>.  Found ${inText.slice(
                        cursor.from,
                        cursor.to,
                    )}.`;
                    errors.push({
                        message,
                        doenetMLrange: {
                            begin: cursor.from + 1,
                            end: cursor.to,
                        },
                    });

                    if (adjustedRange) {
                        element.doenetMLrange = { ...adjustedRange };
                    } else {
                        element.doenetMLrange = {
                            openBegin: tagOpenBegin,
                            openEnd: tagOpenEnd,
                            closeBegin: cursor.from + 1,
                            closeEnd: cursor.to,
                        };
                    }
                    element = {
                        componentType: "_error",
                        props: {},
                        children: [element],
                        state: { message },
                    };
                    break;
                } else if (cursor.name === "MissingCloseTag") {
                    let message = `Invalid DoenetML.  Missing closing tag.  Expected </${tagName}>.`;
                    errors.push({
                        message,
                        doenetMLrange: { begin: cursor.from, end: cursor.to },
                    });

                    if (adjustedRange) {
                        element.doenetMLrange = { ...adjustedRange };
                    } else {
                        element.doenetMLrange = {
                            begin: tagOpenBegin,
                            end: cursor.to,
                        };
                    }
                    element = {
                        componentType: "_error",
                        props: {},
                        children: [element],
                        state: { message },
                    };
                    break;
                } else {
                    // console.log(`error is at position ${cursor.from}, ${cursor.to}`)
                    // console.log(`error part: ${inText.slice(cursor.from, cursor.to)}`)
                    // console.log(`Here is cursor: ${showCursor(cursor)}`)
                    // There are a couple of other things in the entity non-terminal, but nothing of immediate importance
                    if (adjustedTagName !== "_error") {
                        let message = `Invalid DoenetML.  Missing closing tag.  Expected </${tagName}>.`;
                        errors.push({
                            message,
                            doenetMLrange: {
                                begin: cursor.from,
                                end: cursor.to,
                            },
                        });

                        if (adjustedRange) {
                            element.doenetMLrange = { ...adjustedRange };
                        } else {
                            element.doenetMLrange = {
                                begin: tagOpenBegin,
                                end: cursor.to,
                            };
                        }
                        element = {
                            componentType: "_error",
                            props: {},
                            children: [element],
                            state: { message },
                        };
                        break;
                    }
                }
            }
            element.children = mergeConsecutiveStrings(element.children);

            if (adjustedRange) {
                element.doenetMLrange = adjustedRange;
            } else {
                element.doenetMLrange = {
                    openBegin: tagOpenBegin,
                    openEnd: tagOpenEnd,
                    closeBegin: cursor.from + 1,
                    closeEnd: cursor.to,
                };
            }
            return element;
            // @ts-ignore
        } else if (cursor.name === "SelfClosingTag") {
            cursor.firstChild();
            cursor.nextSibling();

            let tagName = inText.substring(cursor.from, cursor.to);
            let adjustedTagName = tagName;

            let tagBegin = cursor.from;

            let message;

            if (tagName[0] === "_") {
                message = `Invalid component type: <${tagName}>.`;
                errors.push({
                    message,
                    doenetMLrange: { begin: tagBegin, end: cursor.to },
                });
                adjustedTagName = "_error";
            }

            let attrs: Element["props"] = {};
            let attrRanges: AttrRange = {};
            while (cursor.nextSibling()) {
                if (cursor.name === "SelfCloseEndTag") {
                    continue;
                }
                //All of the siblings must be Attributes, but we're checking just in case the grammar changes
                if (cursor.name !== "Attribute") {
                    // Note: not sure if can get to this condition.  Errors in self-closing tag
                    // seem to prevent parser from recognizing that it is self-closing.
                    let errorEnd = cursor.to;

                    message = `Invalid DoenetML. Error in self-closing <${tagName}> tag.  Found ${inText.slice(
                        tagBegin - 1,
                        errorEnd,
                    )}`;

                    errors.push({
                        message,
                        doenetMLrange: { begin: tagBegin - 1, end: errorEnd },
                    });
                    adjustedTagName = "_error";

                    break;
                }
                //Attributes always have exactly two children, an AttributeName and an Attribute Value
                //We scrape the content of both from the in string and add them to the attribute array here
                cursor.firstChild();
                let attrName = inText.substring(cursor.from, cursor.to);
                let beginAttributeInd = cursor.from + 1;

                if (cursor.nextSibling() === false) {
                    if (attrName in attrs) {
                        message = `Duplicate attribute ${attrName}.`;
                        errors.push({
                            message,
                            doenetMLrange: {
                                begin: beginAttributeInd,
                                end: cursor.to,
                            },
                        });
                        adjustedTagName = "_error";
                    } else {
                        attrs[attrName] = true;
                    }
                } else {
                    cursor.nextSibling();
                    if (attrName in attrs) {
                        message = `Duplicate attribute ${attrName}.`;
                        errors.push({
                            message,
                            doenetMLrange: {
                                begin: beginAttributeInd,
                                end: cursor.to,
                            },
                        });
                        adjustedTagName = "_error";
                    } else {
                        //fuddling to ignore the quotes
                        let attrValue = inText.substring(
                            cursor.from + 1,
                            cursor.to - 1,
                        );
                        attrs[attrName] = attrValue;
                        attrRanges[attrName] = {
                            attrBegin: beginAttributeInd,
                            attrEnd: cursor.to,
                            begin: cursor.from + 2,
                            end: cursor.to - 1,
                        };
                    }
                }
                //move out of Attribute to maintain loop invariant
                cursor.parent();
            }

            let selfCloseEnd = cursor.to;

            let doenetMLrange = {
                selfCloseBegin: tagBegin,
                selfCloseEnd,
            };

            // console.log(">>>toReturn", {componentType :  tagName, props : attrs, children : []});

            //I have no idea why attrs needs to be destructured
            // but if it isn't, it doesn't work ~50% of the time
            let element: Element = {
                componentType: adjustedTagName,
                props: {},
                children: [],
                doenetMLrange,
            };
            if (adjustedTagName === "_error") {
                element.state = { message: message || "" };
                element.doenetAttributes = {
                    createNameFromComponentType: tagName,
                };
                if (attrs.name) {
                    element.props = { name: attrs.name };
                }
            } else {
                element.props = { ...attrs };
                element.attributeRanges = { ...attrRanges };
            }
            return element;
        } else {
            //Unreachable case, see the grammar for why
            throw Error(
                "Non SelfClosingTag/OpenTag in Element. How did you do that?",
            );
        }
    }
    function compileTopLevel(tc: TreeCursor): Node | undefined {
        if (tc.node.name === "Element") {
            return compileElement(tc.node.cursor());
        } else if (tc.node.name === "Comment") {
            // return a comment that will be ignored,
            // but need it to calculate doenetMLrange
            // from any strings that follow it (needed in particular for macros)
            return {
                componentType: "_comment",
                state: { text: inText.substring(tc.from, tc.to) },
                doenetMLrange: {
                    begin: tc.from,
                    end: tc.to,
                },
            };
        } else if (
            tc.node.name === "Text" ||
            tc.node.name === "Ampersand" ||
            tc.node.name === "EntityReference" ||
            tc.node.name === "CharacterReference"
        ) {
            let txt = inText.substring(tc.node.from, tc.node.to);
            if (txt !== "") {
                return txt;
            }
        } else {
            let message = `Invalid DoenetML.  Found ${inText.substring(
                tc.node.from,
                tc.node.to,
            )}`;
            errors.push({
                message,
                doenetMLrange: { begin: tc.node.from + 1, end: tc.node.to },
            });
            return {
                componentType: "_error",
                props: {},
                children: [],
                state: { message },
                doenetMLrange: { begin: tc.node.from + 1, end: tc.node.to },
            };
        }
    }
    if (!inText) {
        return { components: [], errors };
    }
    let tc = parse(inText);
    let out: Node[] = [];
    if (!tc.firstChild()) {
        return { components: out, errors };
    }
    // console.log("intext",inText)
    // console.log("showCursor",showCursor(tc));

    let first = compileTopLevel(tc);
    if (first !== null && first !== undefined) {
        out.push(first);
    }
    while (tc.nextSibling()) {
        let next = compileTopLevel(tc);
        if (next !== null && next !== undefined) {
            out.push(next);
        }
    }

    out = mergeConsecutiveStrings(out);

    return { components: out, errors };
}

/**
 * pretty-print the tree pointed to by a tree-cursor.
 * Intended for demonstration/debugging
 * @param {TreeCursor} cursor
 * @returns {string}
 */
export function showCursor(cursor: TreeCursor) {
    return showNode(cursor.node);
}

export function showNode(node: SyntaxNode) {
    let str = node.name;
    if (node.firstChild !== null) {
        str += "(" + showNode(node.firstChild) + ")";
    }
    if (node.nextSibling !== null) {
        str += "," + showNode(node.nextSibling);
    }
    return str;
}

// merge consecutive string nodes into one string node
function mergeConsecutiveStrings(nodes: Node[]) {
    let mergedNodes: Node[] = [];
    let prevNode: Node | undefined;
    for (let node of nodes) {
        if (typeof node === "string" && typeof prevNode === "string") {
            prevNode = prevNode + node;
            mergedNodes[mergedNodes.length - 1] = prevNode;
        } else {
            prevNode = node;
            mergedNodes.push(node);
        }
    }

    return mergedNodes;
}
