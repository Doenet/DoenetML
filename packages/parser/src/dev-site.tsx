/*
 * This file is for running a dev test of the codemirror component.
 * It does not show up in the bundled package.
 */

import React from "react";
import ReactDOM from "react-dom/client";
import { CodeMirror } from "@doenet/codemirror";
import { lezerToDast, parse } from "./parser";
import JsonView from "react18-json-view";
import "react18-json-view/src/style.css";
import { filterPositionInfo } from "./dast-to-xml/utils";

const INITIAL_DOENET_SOURCE = `
<p>Use this to test DoenetML</p>
<graph showNavigation="false">

  <line through="(-8,8) (9,6)" />
  <line through="(0,4)" slope="1/2" styleNumber="2" />

  <line equation="y=2x-8" styleNumber="3" />
  <line equation="x=-6" styleNumber="4" />
  
</graph>
`;

ReactDOM.createRoot(document.getElementById("root")!).render(<App />);

function App() {
    const [doenetSource, setDoenetSource] = React.useState(
        INITIAL_DOENET_SOURCE,
    );
    const [dast, setDast] = React.useState({});
    const [omitPosition, setOmitPosition] = React.useState(false);
    const [rawJson, setRawJson] = React.useState(false);

    React.useEffect(() => {
        let dast = lezerToDast(doenetSource);
        if (omitPosition) {
            dast = filterPositionInfo(dast) as any;
        }
        setDast(dast);
    }, [doenetSource, omitPosition]);

    return (
        <div
            style={{ display: "flex", flexDirection: "column", height: "100%" }}
        >
            <div>
                <input
                    type="checkbox"
                    id="include-position"
                    checked={omitPosition}
                    onChange={(e) => {
                        setOmitPosition(e.target.checked);
                    }}
                />
                <label htmlFor="include-position">Omit position</label>
                <input
                    type="checkbox"
                    id="raw-json"
                    checked={rawJson}
                    onChange={(e) => {
                        setRawJson(e.target.checked);
                    }}
                />
                <label htmlFor="raw-json">Show raw JSON</label>
            </div>
            <div
                style={{ display: "flex", overflow: "hidden", height: "100%" }}
            >
                <div style={{ flexBasis: 0, flexGrow: 1, maxWidth: "50%" }}>
                    <CodeMirror
                        onBeforeChange={(val) => {
                            setDoenetSource(val);
                        }}
                        setInternalValueTo={INITIAL_DOENET_SOURCE}
                    />
                </div>
                <div
                    style={{
                        flexBasis: 0,
                        flexGrow: 1,
                        maxWidth: "50%",
                        overflow: "scroll",
                    }}
                >
                    {rawJson ? (
                        <pre>{JSON.stringify(dast, null, 2)}</pre>
                    ) : (
                        <JsonView
                            src={dast}
                            enableClipboard={false}
                            collapsed={(node) => {
                                if (node.indexOrName === "position") {
                                    return true;
                                }
                                return false;
                            }}
                        />
                    )}
                </div>
            </div>
        </div>
    );
}
