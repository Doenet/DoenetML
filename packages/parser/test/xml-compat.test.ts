import { glob } from "glob";
import { describe, expect, it } from "vitest";
import * as fs from "node:fs/promises";
import { parse, showCursor } from "../src/parser";

describe("Xml files parse correctly", async () => {
    const files = glob
        .sync(new URL("./fragments/xml-*.doenet", import.meta.url).pathname)
        .filter((f) => !f.includes("-bad"));
    const fileMap: Record<string, string> = Object.fromEntries(
        files.map((f) => [f.split("/").pop(), f]),
    );

    for (const [filename, filepath] of Object.entries(fileMap)) {
        it(`${filename} parses correctly`, async () => {
            const source = await fs.readFile(filepath, "utf-8");
            const parsed = parse(source);
            expect(showCursor(parsed)).toMatchSnapshot();
        });
    }
});
