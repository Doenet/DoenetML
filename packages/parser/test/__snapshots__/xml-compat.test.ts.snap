// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Xml files parse correctly > xml-1.doenet parses correctly 1`] = `"Document(Element(SelfClosingTag(StartTag,TagName,SelfCloseEndTag)))"`;

exports[`Xml files parse correctly > xml-2.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-3.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)),Element(SelfClosingTag(StartTag,TagName,SelfCloseEndTag)),CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-4.doenet parses correctly 1`] = `"Document(Element(SelfClosingTag(StartTag,TagName,Attribute(AttributeName,Is,AttributeValue),SelfCloseEndTag)))"`;

exports[`Xml files parse correctly > xml-5.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,Attribute(AttributeName,Is,AttributeValue),Attribute(AttributeName,Is,AttributeValue),Attribute(AttributeName,Is,AttributeValue),EndTag),CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-6.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,Attribute(AttributeName,Is,AttributeValue(EntityReference)),EndTag),EntityReference,CharacterReference,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-7.doenet parses correctly 1`] = `"Document(DoctypeDecl,Text,Element(SelfClosingTag(StartTag,TagName,SelfCloseEndTag)))"`;

exports[`Xml files parse correctly > xml-8.doenet parses correctly 1`] = `"Document(ProcessingInst,Element(OpenTag(StartTag,TagName,EndTag),ProcessingInst,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-8b.doenet parses correctly 1`] = `"Document(ProcessingInst)"`;

exports[`Xml files parse correctly > xml-9.doenet parses correctly 1`] = `"Document(Comment,Text,Element(OpenTag(StartTag,TagName,EndTag),Comment,Text,CloseTag(StartCloseTag,TagName,EndTag)),Text,Comment)"`;

exports[`Xml files parse correctly > xml-10.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Cdata,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-12.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,Attribute(AttributeName,Is,AttributeValue),EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-ampersand-1.doenet parses correctly 1`] = `"Document(EntityReference,Text,EntityReference,Text,EntityReference,Text)"`;

exports[`Xml files parse correctly > xml-doenet-ampersand-2.doenet parses correctly 1`] = `"Document(Ampersand,Text,EntityReference,Text,Ampersand,Text)"`;

exports[`Xml files parse correctly > xml-doenet-ampersand-3.doenet parses correctly 1`] = `"Document(Ampersand)"`;

exports[`Xml files parse correctly > xml-doenet-attributes-1.doenet parses correctly 1`] = `"Document(Element(SelfClosingTag(StartTag,TagName,Attribute(AttributeName),Attribute(AttributeName,Is,AttributeValue),Attribute(AttributeName,Is,AttributeValue),SelfCloseEndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-attributes-2.doenet parses correctly 1`] = `"Document(Element(SelfClosingTag(StartTag,TagName,Attribute(AttributeName,Is,AttributeValue),Attribute(AttributeName,Is,AttributeValue(Ampersand)),SelfCloseEndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-lessthan-in-text-1.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-lessthan-in-text-2.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-lessthan-in-text-3.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-lessthan-in-text-4.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-lessthan-in-text-5.doenet parses correctly 1`] = `"Document(Element(OpenTag(StartTag,TagName,EndTag),Text,CloseTag(StartCloseTag,TagName,EndTag)))"`;

exports[`Xml files parse correctly > xml-doenet-no-root.doenet parses correctly 1`] = `"Document(Text,Element(SelfClosingTag(StartTag,TagName,SelfCloseEndTag)),Text,Element(SelfClosingTag(StartTag,TagName,SelfCloseEndTag)),Text)"`;
